source/slang/check.cpp

```diff
@@ -3991,6 +4004,46 @@ namespace Slang
                 else
                 {
                     getSink()->diagnose(expr, Diagnostics::assignNonLValue);
+
+                    // As a special case, check if the LHS expression is derived
+                    // from a `this` parameter (implicitly or explicitly), which
+                    // is immutable. We can give the user a bit more context into
+                    // what is going on.
+                    //
+                    // We will try to handle expressions of the form:
+                    //
+                    //      e ::= "this"
+                    //          | e . name
+                    //          | e [ expr ]
+                    //
+                    // We will unwrap the `e.name` and `e[expr]` cases in a loop.
+                    RefPtr<Expr> e = expr->left;
```

Comment:
A raw pointer should work here, and might be faster. The main reason I used `RefPtr` was actually because the `.As<T>()` method is more convenient and looks nicer than `dynamic_cast`. This would be a candidate for cleanup if we started to get more serious about minimizing the amount of reference-counting operations we do, but I don't see that as the highest priority right now.